### PR TITLE
[#IOPAE-344] Get all user services

### PR DESCRIPTION
**Pull request by giuseppedipinto**
Add **GetServices** API to retrieve all User related services.
The API is paginated.
It's possible to specify optional `limit` and `offset` _queryparam_ to perform specific paginated get operations.

In case the above parameters are not specified, the handler will behave like this:
- `offset` will be 0
- `limit` will assume a default value defined by the config key `PAGINATION_DEFAULT_LIMIT` _(with a default value of 20)_

**Note:**
the value of the `limit` parameter is included in a range that goes from 1 to a maximum value defined by the configuration key `PAGINATION_MAX_LIMIT` _(with a default value of 100)_.

#### List of Changes
- new `GetServices` function with relative handler
- new `apim-filters` utility for building filters for _ApiManagement_ operations
- new `getUserSubscriptions` on _apim-client_
- new `bulkFetch` operation on _fsm store_ for optimize multiple read operations on db
- unit tests

#### How Has This Been Tested?
Unit tests and local environment with production data.

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Review by giamma1295 — APPROVED** [review]
LGTM

File: packages/io-services-cms-models/src/lib/fsm/store.cosmos.ts [hunk 1]
```diff
@@ -55,6 +59,53 @@ export const createCosmosStore = <
...
+                  version: res.eTag,
+                },
+                codec.decode,
+                E.fold(
+                  () => O.none,
+                  (a) => O.some(a)
+                )
+              )
+        )
+      )
```
Comment (rodoviti) [inline packages/io-services-cms-models/src/lib/fsm/store.cosmos.ts:106]: nitpick: IMHO this way it's more readable 😉
```suggestion
      TE.map((operationResponses) =>
        operationResponses.map(
          flow(
            O.fromPredicate((res) => res.statusCode === 404),
            // if present, try to decode in the expected shape
            O.chain((res) =>
              pipe(
                {
                  ...res.resourceBody,
                  last_update: res.resourceBody
                    ? new Date(
                        // eslint-disable-next-line no-underscore-dangle
                        (res.resourceBody._ts as number) * 1000
                      ).toISOString() // Unix timestamp
                    : new Date().toISOString(),
                  version: res.eTag,
                },
                codec.decode,
                E.fold(() => O.none, O.some)
              )
            )
          )
        )
      )
```

**Review by rodoviti — APPROVED** [review]
see suggestions 😉

File: apps/io-services-cms-webapp/src/webservice/controllers/get-services.ts [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+) =>
+  subscriptions
+    // associate each service with its subscription
+    .map((subscription) => ({
+      subscription,
+      service: services.find((s) => s.id === subscription.name),
+    }))
+    // be sure both subscription and service are not undefined
+    .filter(
+      (_): _ is ServiceSubscriptionPair =>
```
Comment (gunzip) [inline apps/io-services-cms-webapp/src/webservice/controllers/get-services.ts:164]: I'd extract this into `isServiceSubscriptionPair`

File: apps/io-services-cms-webapp/src/utils/apim-filters.ts [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+  E.isLeft(filterManipulation)
+    ? O.none
+    : pipe(
+        filterManipulation.right,
+        fnComposeFilter,
+        fnInverseFilter,
+        fnCheckFilterType,
+        fnInverseFilterCloseBrackets,
+        fnEvaluateFilterResult
+      );
```
[the remark this answers is not in the document]
Comment (giuseppedipinto) [reply apps/io-services-cms-webapp/src/utils/apim-filters.ts:117]: @rodoviti as @gunzip said, `E.bimap` becomes `E.fold`

File: apps/io-services-cms-webapp/src/webservice/controllers/get-services.ts [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+  pipe(
+    obj,
+    t.type({ id: NonEmptyString }).decode,
+    TE.fromEither,
+    TE.mapLeft(
+      (err) =>
+        new Error(`Cannot decode object to get id, ${readableReport(err)}`)
+    ),
+    TE.map((_) => parseOwnerIdFullPath(_.id))
+  );
```
[the remark this answers is not in the document]
Comment (giuseppedipinto) [reply apps/io-services-cms-webapp/src/webservice/controllers/get-services.ts:85]: IMO this is the best mix:
```suggestion
const pickId = (obj: unknown): E.Either<Error, NonEmptyString> =>
  pipe(
    obj,
    t.type({ id: NonEmptyString }).decode,
    E.bimap(
      (err) =>
        new Error(`Cannot decode object to get id, ${readableReport(err)}`),
      ({ id }) => parseOwnerIdFullPath(id)
    )
  );
```

File: apps/io-services-cms-webapp/src/webservice/controllers/get-services.ts [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+          : ({} as ServiceResponsePayload)
+      )
+    )
+  );
+
+const getLimit = (limit: O.Option<number>, defaultValue: number) =>
+  O.isSome(limit) ? limit.value : defaultValue;
+
+const getOffset = (offset: O.Option<number>) =>
+  O.isSome(offset) ? offset.value : 0;
```
[the remark this answers is not in the document]
Comment (giuseppedipinto) [reply apps/io-services-cms-webapp/src/webservice/controllers/get-services.ts:144]:
```suggestion
  pipe(
    offset,
    O.getOrElse(() => 0)
  );
```

File: apps/io-services-cms-webapp/src/utils/apim-filters.ts [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+/** Build `FilterManipulationResultType` object */
+const fnBuildFilterManipulation = (
+  decodedApimFilter: E.Either<t.Errors, ApimFilterType>
+): E.Either<Error, FilterManipulationResultType> =>
+  E.isLeft(decodedApimFilter)
+    ? E.left(new Error("error"))
+    : E.right({
+        filter: decodedApimFilter.right,
+        result: "",
+      });
```
[the remark this answers is not in the document]
Comment (giuseppedipinto) [reply apps/io-services-cms-webapp/src/utils/apim-filters.ts:102]:
```suggestion
  pipe(
    decodedApimFilter,
    E.bimap(flow(readableReport, E.toError), (filter) => ({
      filter,
      result: "",
    }))
  );
```

**Comment by giuseppedipinto** [comment]
@giamma1295 @rodoviti @gunzip thanks for your suggestions, I appreciate it!
See 7b6742b for improvements.

**Pull request merged**